### PR TITLE
Update concat_ws scalar function to support Utf8View

### LINKED ISSUE #11837: Update `concat_ws` scalar function to support `Utf8View`
Part of https://github.com/apache/datafusion/issues/11752 and https://github.com/apache/datafusion/issues/11790

Currently, a call to `CONCAT_WS` with a Utf8View datatypes induces a cast. After the change that fixes this issue, it should not.

```sql
query TT
EXPLAIN SELECT
  concat_ws(', ', column1_utf8view, column2_utf8view) as c
FROM test;
----
logical_plan
01)Projection: concat_ws(Utf8(", "), CAST(test.column1_utf8view AS Utf8), CAST(test.column2_utf8view AS Utf8)) AS c
02)--TableScan: test projection=[column1_utf8view, column2_utf8view]
```

concat_ws is defined here: https://github.com/apache/datafusion/blob/main/datafusion/functions/src/string/concat_ws.rs

### Is your feature request related to a problem or challenge?

We are working to add complete StringView support in DataFusion, which permits potentially much faster processing of string data. See https://github.com/apache/datafusion/issues/10918 for more background.

Today, most DataFusion string functions support DataType::Utf8 and DataType::LargeUtf8 and when called with a StringView argument DataFusion will cast the argument back to DataType::Utf8 which is expensive.

To realize the full speed of StringView, we need to ensure that all string functions support the DataType::Utf8View directly.

### Describe the solution you'd like

Update the function to support DataType::Utf8View directly

### Describe alternatives you've considered

The typical steps are:
1. Write some tests showing the function doesn't support Utf8View (see the tests in [`string_view.slt`](https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/string_view.slt) to ensure the arguments are not being cast
2. Change the `Signature` of the function to accept `Utf8View` in addition to `Utf8`/`LargeUtf8`
3. Update the implementation of the function to operate on `Utf8View`

Example PRs
* Update to use an arrow kernel that already supports StringView:  https://github.com/apache/datafusion/pull/11787
* Change the implementation to support StringView directly: https://github.com/apache/datafusion/pull/11676
* Change implementation (option 2): https://github.com/apache/datafusion/pull/11556

### Additional context

The documentation of string functions can be found here: https://datafusion.apache.org/user-guide/sql/scalar_functions.html#string-functions

To test a function with StringView with `datafusion-cli` you can use an example like this (replacing `starts_with` with the relevant function)
```sql
> create table foo as values (arrow_cast('foo', 'Utf8View'), arrow_cast('bar', 'Utf8View'));
0 row(s) fetched.
Elapsed 0.043 seconds.

> select starts_with(column1, column2) from foo;
+--------------------------------------+
| starts_with(foo.column1,foo.column2) |
+--------------------------------------+
| false                                |
+--------------------------------------+
1 row(s) fetched.
Elapsed 0.015 seconds.
```

To see if it is using utf8 view, use `EXPLAIN` to see the plan and verify there is no `CAST`. In this example the `CAST(column1@0 AS Utf8)` indicates that the function is not using `Utf8View` natively

```sql
> explain select starts_with(column1, column2) from foo;
+---------------+------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                         |
+---------------+------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: starts_with(CAST(foo.column1 AS Utf8), CAST(foo.column2 AS Utf8))                                                |
|               |   TableScan: foo projection=[column1, column2]                                                                               |
| physical_plan | ProjectionExec: expr=[starts_with(CAST(column1@0 AS Utf8), CAST(column2@1 AS Utf8)) as starts_with(foo.column1,foo.column2)] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                                                              |
|               |                                                                                                                              |
+---------------+------------------------------------------------------------------------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.006 seconds.
```

It is also often good to test with a constant as well (likewise there should be no cast):

```sql
> explain select starts_with(column1, 'foo') from foo;
+---------------+----------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                     |
+---------------+----------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: starts_with(CAST(foo.column1 AS Utf8), Utf8("foo"))                                          |
|               |   TableScan: foo projection=[column1]                                                                    |
| physical_plan | ProjectionExec: expr=[starts_with(CAST(column1@0 AS Utf8), foo) as starts_with(foo.column1,Utf8("foo"))] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                                          |
|               |                                                                                                          |
+---------------+----------------------------------------------------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.002 seconds.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by devanbenz**
## Which issue does this PR close?

Closes #11837

### CODE REVIEW

**Comment by alamb** [comment]
Marking as draft as it looks like we are still working out the CI checks

File: datafusion/functions/src/string/concat_ws.rs [hunk 1]
```diff
@@ -46,9 +45,8 @@ impl Default for ConcatWsFunc {
 
 impl ConcatWsFunc {
     pub fn new() -> Self {
-        use DataType::*;
         Self {
-            signature: Signature::variadic(vec![Utf8], Volatility::Immutable),
+            signature: Signature::variadic_any(Volatility::Immutable),
```
Comment (devanbenz) [inline datafusion/functions/src/string/concat_ws.rs:49]: @alamb having a few issues with the `signature` here. I'm assuming that `variadic` will allow for a vector of common types but it effectively will coerce the other types in to whatever the first type is? In this case the separator is generally always a `Utf8` type which then causes the subsequent parameters to be cast to a utf8.

```
    pub fn new() -> Self {
        use DataType::*;
        Self {
            signature: Signature::variadic(
                vec![Utf8, Utf8View, LargeUtf8],
                Volatility::Immutable,
            ),
        }
    }
```

Is there something similar to: 

`Exact(Utf8, variadic([utf8, utf8view, largeutf8]))` ?

Looking through the type signatures I don't see an established way to do that (that I'm aware of). Any suggestions?

Comment (alamb) [reply datafusion/functions/src/string/concat_ws.rs:49]: > Looking through the type signatures I don't see an established way to do that (that I'm aware of). Any suggestions?

One way to do this is to explicitly list all the valid combinations using [`TypeSignature::OneOf`](https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.TypeSignature.html#variant.OneOf)

for example, 
https://github.com/apache/datafusion/blob/e8ac93a0ffd490bd549872366bf7fd1b77ffd7f8/datafusion/functions/src/string/btrim.rs#L56-L62

File: datafusion/functions/src/string/concat_ws.rs [hunk 2]
```diff
@@ -155,21 +169,53 @@ impl ScalarUDFImpl for ConcatWsFunc {
...
+                                ColumnarValueRef::NullableLargeStringArray(string_array)
+                            } else {
+                                ColumnarValueRef::NonNullableLargeStringArray(string_array)
+                            };
+                            columns.push(column);
+                        },
+                        DataType::Utf8View => {
+                            let string_array = as_string_view_array(array)?;
+
+                            data_size += string_array.len();
```
Comment (2010YOUY01) [inline datafusion/functions/src/string/concat_ws.rs:207]: `data_size` is the estimated string array size after concatenation, I think here it should increment sum of inner buffers' size instead

(And everything else LGTM)

Comment (devanbenz) [reply datafusion/functions/src/string/concat_ws.rs:207]: Using `data_size += string_array.data_buffers().len();` ?

**Review by 2010YOUY01 — APPROVED** [review]
Awesome! Thank you

File: datafusion/sqllogictest/test_files/string_view.slt [hunk 3]
```diff
@@ -796,7 +796,7 @@ EXPLAIN SELECT
 FROM test;
 ----
 logical_plan
-01)Projection: concat_ws(Utf8(", "), CAST(test.column1_utf8view AS Utf8), CAST(test.column2_utf8view AS Utf8)) AS c
+01)Projection: concat_ws(Utf8(", "), test.column1_utf8view, test.column2_utf8view) AS c
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/string_view.slt:799]: ❤️

**Review by alamb — APPROVED** [review]
Thank you @devanbenz  -- this looks great to me

Thank you @2010YOUY01  for the review

**Comment by alamb** [comment]
Thanks again @devanbenz and @2010YOUY01

**Pull request merged**